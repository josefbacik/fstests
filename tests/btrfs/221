#! /bin/bash
# SPDX-License-Identifier: GPL-2.0
# Copyright (c) 2020 Oracle. All Rights Reserved.
#
# FS QA Test 221
#
# Test for seed device-delete on a sprouted FS.
# Requires kernel patch
#    btrfs: fix put of uninitialized kobject after seed device delete
#
# Steps:
#  Create a seed FS. Add a RW device to make it sprout FS and then delete
#  the seed device.

seq=`basename $0`
seqres=$RESULT_DIR/$seq
echo "QA output created by $seq"

here=`pwd`
tmp=/tmp/$$
status=1	# failure is the default!
trap "_cleanup; exit \$status" 0 1 2 3 15

_cleanup()
{
	cd /
	rm -f $tmp.*
}

# get standard environment, filters and checks
. ./common/rc
. ./common/filter

# remove previous $seqres.full before test
rm -f $seqres.full

# real QA test starts here

# Modify as appropriate.
_supported_fs generic
_supported_os Linux
_require_test
_require_scratch_dev_pool 2

_scratch_dev_pool_get 2

seed=$(echo $SCRATCH_DEV_POOL | awk '{print $1}')
sprout=$(echo $SCRATCH_DEV_POOL | awk '{print $2}')

_mkfs_dev $seed
_mount $seed $SCRATCH_MNT

$XFS_IO_PROG -f -d -c "pwrite -S 0xab 0 1M" $SCRATCH_MNT/foo > /dev/null
_scratch_unmount
$BTRFS_TUNE_PROG -S 1 $seed

# Mount the seed device and add the rw device
_mount -o ro $seed $SCRATCH_MNT
$BTRFS_UTIL_PROG device add -f $sprout $SCRATCH_MNT
_scratch_unmount

# Now remount
_mount $sprout $SCRATCH_MNT
$XFS_IO_PROG -f -d -c "pwrite -S 0xcd 0 1M" $SCRATCH_MNT/bar > /dev/null

echo --- before delete ----
od -x $SCRATCH_MNT/foo
od -x $SCRATCH_MNT/bar

$BTRFS_UTIL_PROG device delete $seed $SCRATCH_MNT
_scratch_unmount
_btrfs_forget_or_module_reload
_mount $sprout $SCRATCH_MNT

echo --- after delete ----
od -x $SCRATCH_MNT/foo
od -x $SCRATCH_MNT/bar

_scratch_dev_pool_put

# success, all done
status=0
exit
