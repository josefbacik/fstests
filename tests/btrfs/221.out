QA output created by 221
--- before delete ----
0000000 abab abab abab abab abab abab abab abab
*
4000000
0000000 cdcd cdcd cdcd cdcd cdcd cdcd cdcd cdcd
*
4000000
--- after delete ----
0000000 abab abab abab abab abab abab abab abab
*
4000000
0000000 cdcd cdcd cdcd cdcd cdcd cdcd cdcd cdcd
*
4000000
